readme.md

```diff
@@ -47,126 +62,497 @@ This package exports [ESM](https://gist.github.com/sindresorhus/a39789f98801d908
+
+3. Type Assertion Methods:
+   - `assertSystem`, `assertUser`, `assertAssistant`, `assertRefusal`, `assertFuncCall`, `assertFuncResult`, `assertToolCall`, `assertToolResult`
+
+4. Conversion Method:
+   - `fromChatMessage`: Converts an `openai-fetch` `ChatMessage` to a Dexter `Msg` type
+
+### Telemetry
+
+Dexter includes a telemetry system for tracking and logging model operations. The telemetry system is based on the OpenTelemetry standard and can be integrated with various observability platforms.
+
+1. **Default Telemetry**: By default, Dexter uses a no-op telemetry provider that doesn't perform any actual logging or tracing.
+
+2. **Custom Telemetry**: You can provide your own telemetry provider when initializing models. The provider should implement the `Telemetry.Provider` interface:
+
+   ```typescript
+   interface Provider {
+     startSpan<T>(options: SpanOptions, callback: (span: Span) => T): T;
+     setTags(tags: { [key: string]: Primitive }): void;
+   }
+   ```
+
+3. **Span Attributes**: Dexter automatically adds various attributes to telemetry spans, including model type, provider, input tokens, output tokens, and more.
+
+4. **Usage**: Telemetry is used internally in the `AbstractModel` class to wrap the `run` method, providing insights into model execution.
+
+### Caching
+
+Dexter provides a flexible caching system to improve performance and reduce API calls:
+
+1. **Cache Interface**: The cache must implement the `CacheStorage` interface:
+
+   ```typescript
+   interface CacheStorage<KeyType, ValueType> {
+     get: (key: KeyType) => Promise<ValueType | undefined> | ValueType | undefined;
+     set: (key: KeyType, value: ValueType) => Promise<unknown> | unknown;
+   }
+   ```
+
+2. **Default Cache Key**: Dexter uses a default cache key function that creates a SHA512 hash of the input parameters.
+
+3. **Custom Cache**: You can provide your own cache implementation when initializing models:
+
+   ```typescript
+   import { ChatModel } from '@dexaai/dexter';
+
+   const customCache = new Map();
+   const chatModel = new ChatModel({ cache: customCache });
+   ```
+
+4. **Cache Usage**: Caching is automatically applied in the `AbstractModel` class. Before making an API call, it checks the cache for a stored response. After receiving a response, it stores it in the cache for future use.
+
+5. **Cache Invalidation**: Cache invalidation is left to the user. You can implement your own cache invalidation strategy based on your specific use case.
+
+### Tokenization
+
+Dexter includes a tokenization system based on the `tiktoken` library, which is used by OpenAI for their models. This system is crucial for accurately counting tokens and managing model inputs and outputs.
+
+1. **Tokenizer Creation**: The `createTokenizer` function creates a `Tokenizer` instance for a specific model:
+
+   ```typescript
+   const tokenizer = createTokenizer('gpt-3.5-turbo');
```

Comment:
the multiple default imports mean that as a dev reading this, even knowing dexter pretty well, I have no idea where I'd import `createTokenizer` from, for instance.

Reply:
good thing your ide does if you type `createTo<complete>`